discovery/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,266 @@
+	return fmt.Sprintf("%s/%s", conn.parentPath(service), service.ID)
+}
+
+func (conn *ZooKeeper) getAppKey(appName string) string {
+	return fmt.Sprintf("%s/%s", conn.Prefix, appName)
+}
+
+var zookeeperUpstreams = make(map[string][]ServiceNode)
+
+// CheckForUpstreamChanges checks another zookeeper node for changes
+func (conn *ZooKeeper) CheckForUpstreamChanges(backendName, backendTag string) bool {
+	// TODO: is there a way to filter by tag in zookeeper?
+	services, err := conn.getServices(backendName)
+	if err != nil {
+		log.Errorf("Failed to query %v: %s", backendName, err)
+		return false
+	}
+	didChange := zookeeperCompareForChange(zookeeperUpstreams[backendName], services)
+	if didChange || len(services) == 0 {
+		// We don't want to cause an onChange event the first time we read-in
+		// but we do want to make sure we've written the key for this map
+		zookeeperUpstreams[backendName] = services
+	}
+	return didChange
+}
+
+func (conn *ZooKeeper) getServices(appName string) ([]ServiceNode, error) {
+	var services []ServiceNode
+
+	key := conn.getAppKey(appName)
+	children, _, error := conn.Client.Children(key)
+	if error != nil {
+		return services, error
+	}
+	for i := range children {
+		path := fmt.Sprintf("%s/%s", key, children[i])
+		data, _, error := conn.Client.Get(path)
+		if error != nil {
+			return services, error
+		}
+		srv, error := decodeZooKeeperNodeValue(data)
+		if error != nil {
+			return services, error
+		}
+		services = append(services, srv)
+	}
+	return services, nil
+}
+
+// Compare the two arrays to see if the address or port has changed
+// or if we've added or removed entries.
+func zookeeperCompareForChange(existing, new []ServiceNode) (changed bool) {
+	if len(existing) != len(new) {
+		return true
+	}
+
+	sort.Sort(ByZooKeeperServiceID(existing))
+	sort.Sort(ByZooKeeperServiceID(new))
+	for i, ex := range existing {
+		if ex.Address != new[i].Address ||
+			ex.Port != new[i].Port {
+			return true
+		}
+	}
+	return false
+}
+
+func (conn ZooKeeper) createParentPath(path string) error {
+	pathElements := strings.Split(path, "/")[1:]
+	sep := "/"
+	newPath := ""
+	for i := range pathElements {
+		newPath = strings.Join([]string{newPath, sep, pathElements[i]}, "")
+		if _, err := conn.Client.Create(
+			newPath,
+			nil,
+			0,
+			zk.WorldACL(zk.PermAll)); err != nil && err != zk.ErrNodeExists {
+			return err
+		}
+	}
+	return nil
+}
+
+func (conn ZooKeeper) registerService(service *discovery.ServiceDefinition) error {
+	key := conn.getNodeKey(service)
+	if err := conn.createParentPath(conn.parentPath(service)); err != nil {
+		return err
+	}
+	value := encodeZooKeeperNodeValue(service)
+	if _, err := conn.Client.Create(
+		key,
+		[]byte(value),
+		zk.FlagEphemeral,
+		zk.WorldACL(zk.PermAll)); err != nil && err != zk.ErrNodeExists {
+		return err
+	}
+	// Set the watcher and trigger the call via `Set`
+	_, _, _, err := conn.Client.GetW(key)
```

Comment:
I'm still really unclear on what we're trying to do with this. It reads to me like:

- we create a connection with an event callback in `NewZooKeeperConfig`
- we create a watch on a key and throw away the channel here in `registerService`
- we set a value on that key here in `registerService`
- we start a goroutine here in `registerService`
- when an event callback that goroutine checks the service key, creates a new watch and throws away that new channel. If there's an error in doing so it deregisters.

What is sending messages to the event callback in the first place? Why are we creating watches and throwing them away by doing `GetW` in the `ttlHandler` instead of doing `Get`?